### PR TITLE
rest.Config: support configuring an explict proxy URL

### PR DESCRIPTION
**Pull request by tebeco**
### EDIT:
I'll be reading the 7 mails received by tomorrow to understand properly how PR and interaction should be done in the repo ^^
for example ... I copy/pasted bot interaction with the PR description, probably should have removed that first

##
Manual rebase of a 9month PR : https://github.com/kubernetes/kubernetes/pull/81443

This PR is a "re-fork" of `https://github.com/kubernetes/kubernetes` on `master` then rebased `https://github.com/mikedanese/kubernetes/tree/socks5`
and manually fixed conflict

As the original PR seems to have some "attraction", I tried to helped with what I could.
If this is considered "Forbidden" to do so, then, my apologies

## Original Description

With support of http, https, and socks5 proxy support. We already
support configuring this via environmnet variables, but this approach
becomes inconvenient dealing with multiple clusters on different
networks, that require different proxies to connect to. Most solutions
require wrapping clients (like kubectl) in bash scripts.

Part of: kubernetes/client-go#351

/kind feature

```release-note
rest.Config now supports a flag to override proxy configuration that was previously only configurable through environment variables.
```
cc @dlipovetsky @kubernetes/sig-api-machinery-pr-reviews

### CODE REVIEW

**Comment by tebeco** [comment]
All credits goes to @mikedanese,

just hope this won't cause a bug in the "machinery" here that it detects the it CLA (also probably because he's the Author of the commit)

I signed the CLA a bit after submitting the PR
image

**Comment by tebeco** [comment]
image

welllllll he was the owner of the previous one here :
https://github.com/kubernetes/kubernetes/pull/81443

now that's weird ^^

**Comment by tebeco** [comment]
also not sure if you can tell me @lavalamp if i missed anything else in the PR ?
It's a way more complex than other repo so i'm not sure everything is ready

i see that i need someone to run `/ok-to-test`, but not sure that would be me, or if i need something special before it could be ran

**Comment by lavalamp** [comment]
Mike has the latest state on this. I probably won't be able to find time to
review this. I pinged Mike internally to let us know what the state is on
this.

**Comment by tebeco** [comment]
thx a lot for that ;)

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -131,6 +131,13 @@ type Config struct {
 	// Version forces a specific version to be used (if registered)
 	// Do we need this?
 	// Version string
+
+	// ProxyURL is the URL to the proxy to be used for all requests made by this
+	// client. URLs with "http", "https", and "socks5" schemes are supported.
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/rest/config.go:136]: socks5 currently won't work for exec/attach/log streaming commands. Have you confirmed that these work over http/https proxy?

**Comment by mikedanese** [comment]
Does this need to get plumbed through kubeconfig to be useful? Or do we want this in rest.Config as step 1, then implement kubeconfig support in a followup?

**Comment by tebeco** [comment]
to be fair @mikedanese this PR was just a "duplicate" of one of yours here :
https://github.com/kubernetes/kubernetes/pull/81443

Also i did not pay enough attention to see you were a maintainer here.
Maybe i was wrong but i understood that ideally that would help us to set a proxy in the `.kube/config` file

Why ?
Well because when we use Managed cluster like AKS / GKS ... the network traffic cannot reach the API unless we use a very specific proxy
That proxy would also be specific to that cluster, so yeah, that's per Cluster and in the Config file

File: staging/src/k8s.io/client-go/rest/config.go (hunk 1 above)
Comment (tebeco) [reply staging/src/k8s.io/client-go/rest/config.go:136]: as of today, we have to create custom alias like:
`kubeclt` => default binary / no proxy
`kdev` => set `http_proxy` for dev / forward to `kubectl` binary / unset `http_proxy` (not to polute the shell)
`kprd` ....

That's kind of a mess but `attach/exec/log` seems to work

**Comment by mikedanese** [comment]
Totally appreciate the help. I can plumb this through to kubeconfig in the old PR without much work. Supporting streaming endpoints over socks5 proxy is more involved though, however it seems like this would be useful for many people using standard http proxies so it might be worth it to move forward as is and fix socks5 later.

**Comment by tebeco** [comment]
> it might be worth it to move forward as is and fix socks5 later.

yup, that would also let one of the 2 PR to be "minimal" and land it in one of the next release and so on
I have no idea of the ratio of people/company using `HTTP vs SOCKS` proxy TBF

So you prefer i close this PR if you take back #81443 ?
I made sure this is enable just in case
image

**Comment by tebeco** [comment]
@mikedanese 
closing i suppose :D ?

**Pull request closed without merging**